

#Package Installation
if (!requireNamespace("BiocManager", quietly = TRUE))
  install.packages("BiocManager")

BiocManager::install("ROntoTools")

###########################################################################################################################################################
#Basically a library call
require(graph)
require(ROntoTools)

#Download and parse the pathways available in KEGG, this depends on your connection, but it doesn't take too long
#Pathways are downloaded based on the 'organism' parameter, in this case, it is "hsa" which represents human pathways
#The verbose parameter would allow us to view the progress, I choose to enable it, however one should be wary of enabling prints on servers
kpg <- keggPathwayGraphs("hsa",verbose = TRUE)

###########################################################################################################################################################
#The above code loaded the cache, so it should only be run once!
#If there is a need to update the package, this is the command that needs to be run, this one can also be time consuming
#This command took roughly 5 minutes at 80.20 download Mbps and 94.47 Upload Mbps


#At the end of the command, there was a permission denied error
#cannot remove file 'C:\Users\yohan\AppData\Local\Temp\RtmpeSahyJ\ROntoTools53144b973f36', reason 'Permission denied'
#This folder was empty, the command must have been able to empty it, simply not delete the folder itself.

kpg <- keggPathwayGraphs("hsa", updateCache = TRUE, verbose = TRUE)

###########################################################################################################################################################

#This command just shows that the object kpg, generated by the two functions above, is a large list of pathways, each indicated like so

#This may mean that we will require a conversion list at some point to translate the results... maybe the package does it on its own
#Note: head prints the first few elements of an object, not all of them
head(names(kpg))
#Result
#[1] "path:hsa01521" "path:hsa01522" "path:hsa01523" "path:hsa01524" "path:hsa03008" "path:hsa03013"

###########################################################################################################################################################

#This command allows inspection of a specific path
#In this case, it is the mitotic cell cycle for humans (of course)
#Here is the website 
#https://www.genome.jp/dbget-bin/www_bget?pathway+hsa04110
#This result gives the nuber of nodes and edges

kpg[["path:hsa04110"]]
#Result
#A graphNEL graph with directed edges
#Number of Nodes = 124 
#Number of Edges = 632

###########################################################################################################################################################

#To better understand the Nodes and Graphs, we can use the following two functions

#This function prints the first few nodes
head(nodes(kpg[["path:hsa04110"]]))
#Result
#[1] "hsa:1029"  "hsa:51343" "hsa:4171"  "hsa:4172"  "hsa:4173"  "hsa:4174" 

#This function prints the first few edges
head(edges(kpg[["path:hsa04110"]]))
#Result
#$`hsa:1029`
#[1] "hsa:4193" "hsa:1019" "hsa:1021" "hsa:595"  "hsa:894"  "hsa:896" 
# 
#$`hsa:51343`
#[1] "hsa:983"   "hsa:85417" "hsa:891"   "hsa:9133" 
# 
#$`hsa:4171`
#character(0)
# 
#$`hsa:4172`
#character(0)
# 
#$`hsa:4173`
#character(0)
# 
#$`hsa:4174`
#character(0)


#My understanding of nodes and edges is a little fuzzy, however I think they are best explaiend in the following way
#Nodes --> representation of biomolecules, such as proteins, genes and metabolites
#Edges --> representation of the types of associations between two nodes, such as physical interactions and co-expression of mRNAs
#Thus it appears that calling 'edges' actually gives the other nodes that a specific node connects to.
#To obtain the exact type of interaction, we do the following function

#By declaring the attr as 'subtype' we are simply asking the specific interaction, in this occurence, the examples are all of inhibiton
head(edgeData(kpg[["path:hsa04110"]], attr = "subtype"))
#Result
#$`hsa:1029|hsa:4193`
#[1] "inhibition"
# 
#$`hsa:1029|hsa:1019`
#[1] "inhibition"
# 
#$`hsa:1029|hsa:1021`
#[1] "inhibition"
# 
#$`hsa:1029|hsa:595`
#[1] "inhibition"
# 
#$`hsa:1029|hsa:894`
#[1] "inhibition"
# 
#$`hsa:1029|hsa:896`
#[1] "inhibition"
###########################################################################################################################################################

#This function sets the same weight to all the interactions of the same type


#Basically, no weights are ever attributed to pathways, it will be up to us to figure out which weight distribution to use
#The details of its utility are still a little fuzzy, but it shouldn't be too hard to figure out.
kpg <- setEdgeWeights(kpg, edgeTypeAttr = "subtype",edgeWeightByType = list(activation = 1, inhibition = -1,expression = 1, repression = -1),defaultWeight = 0)
#Results
#kpg now contains a list of grpahs with weighted edges


#This simply shows us that weights have been given as intended.
#Note that if we call this action before the previous function (setEdgeWeights) the "weight" attribute would be unknown and nothing would be returned
head(edgeData(kpg[["path:hsa04110"]], attr = "weight"))
#Result
#$`hsa:1029|hsa:4193`
#[1] -1
# 
#$`hsa:1029|hsa:1019`
#[1] -1
# 
#$`hsa:1029|hsa:1021`
#[1] -1
# 
#$`hsa:1029|hsa:595`
#[1] -1
# 
#$`hsa:1029|hsa:894`
#[1] -1
# 
#$`hsa:1029|hsa:896`
#[1] -1

###########################################################################################################################################################

#This function stores the title of pathways (not just the ids) into the kpn list
kpn <- keggPathwayNames("hsa")

#This prints the first few elements of that list
head(kpn)
#Results
#path:hsa00010                                  path:hsa00020                              path:hsa00030 
#"Glycolysis / Gluconeogenesis"                 "Citrate cycle (TCA cycle)"                "Pentose phosphate pathway" 
#path:hsa00040                                  path:hsa00051                              path:hsa00052 
#"Pentose and glucuronate interconversions"     "Fructose and mannose metabolism"           "Galactose metabolism" 




###########################################################################################################################################################


#HERE STARTS THE EXAMPLE

#This is PRE-PROCESSED data set from ArrayExpress that studies the peripheral blood mononuclear cells (PBMC) which is real cool, but not important
#What is important is that the data is pre-processed.
#Here that data was pre-processed using the 'limma' R package, it may be wise to do the same, however I can obtain similar results using DESeq2 (package de mon stage)

#IMPORTANT, THESE HAVE ALREADY BEEN SORTED BY P-VALUE!!!!!
load(system.file("extdata/E-GEOD-21942.topTable.RData", package = "ROntoTools"))

#Print the first few items of the generated example, it has automatically been stored in 'top'
head(top)
#Results
#                 logFC      P.Value    adj.P.Val     entrez
#200946_x_at -1.0175141 5.833411e-13 4.172652e-09   hsa:2746
#228697_at   -3.6479368 7.985427e-13 4.172652e-09 hsa:135114
#210254_at    3.2807123 3.086572e-12 9.677020e-09    hsa:932
#234726_s_at -0.9792301 7.368175e-12 1.760593e-08  hsa:64418
#215905_s_at -1.7733135 7.861797e-12 1.760593e-08   hsa:9410
#235542_at   -0.9447467 1.617944e-11 2.536288e-08 hsa:200424


#This is how we would select differentially expressed genes at 1% and save their fold change in a vector
#fc and theirp-values in a vector 'pv'
fc <- top$logFC[top$adj.P.Val <= .01]
names(fc) <- top$entrez[top$adj.P.Val <= .01]
head(fc)
#Result
#  hsa:2746 hsa:135114    hsa:932  hsa:64418   hsa:9410 hsa:200424 
#-1.0175141 -3.6479368  3.2807123 -0.9792301 -1.7733135 -0.9447467 


#Same thing but with p value
pv <- top$P.Value[top$adj.P.Val <= .01]
names(pv) <- top$entrez[top$adj.P.Val <= .01]
head(pv)
#Result
#    hsa:2746   hsa:135114      hsa:932    hsa:64418     hsa:9410   hsa:200424 
#5.833411e-13 7.985427e-13 3.086572e-12 7.368175e-12 7.861797e-12 1.617944e-11



#Alternatively to the two functions above, we can do an analysis on all genes and store them in a variable
#Here we store the log Fold change in fcALL
fcAll <- top$logFC
#Declare the names of fcALL
names(fcAll) <- top$entrez

#Same as the two above, but with p-value
pvAll <- top$P.Value
names(pvAll) <- top$entrez


#This contains all the genes measured in the analysis
#I can't tell how it uses the above 2 functions
ref <- top$entrez
head(ref)
#Results
#[1] "hsa:2746"   "hsa:135114" "hsa:932"    "hsa:64418"  "hsa:9410"   "hsa:200424"


###########################################################################################################################################################

#About node weights
#Node weights are used to encode for the significance of each gene, the term described as 'alpha' in (article: Proceedings of the International 
#Conference on Machine Learning Applications (ICMLA),). There are two alternative formulas to incorporate gene significance.
#these are alpha1MR and alphaMLG
#To set the weight using these calculations, use the following function
#Here we set the weights using the alphaMLG formula in accordance with the pv (selection of differentially expressed genes of significance 1% (via p-value))
kpg <- setNodeWeights(kpg, weights = alphaMLG(pv), defaultWeight = 1)

#Print the first few edges of node hsa04110
head(nodeWeights(kpg[["path:hsa04110"]]))
#Result
# hsa:1029 hsa:51343  hsa:4171  hsa:4172  hsa:4173  hsa:4174 
#1.0000000 1.0000000 0.8120949 1.0000000 1.0000000 1.0000000


###########################################################################################################################################################
#At this point, all the elements of the analysis have been performed (of course we can remove the prints)
#The critical parts are kpg, fc, and ref

#The pe function is called to perform the analysis, the accuracy is determined by nboot (bigger=more accurate)
#nboot significes number of bootstrap iterations
#This is basically the number of times a statistical test is performed with random sampling replacement, the more it is performed, the more accurate it gets
#Note: verbose should be set to false in final product
peRes <- pe(x = fc, graphs = kpg, ref = ref, nboot = 200, verbose = TRUE)
#Took 10 ish seconds

#Show the first few results
head(Summary(peRes))
#Results
#               totalAcc totalPert totalAccNorm totalPertNorm       pPert        pAcc         pORA        pComb  pPert.fdr   pAcc.fdr    pORA.fdr    pComb.fdr
#path:hsa05010  17.90716 129.19653    0.4766589      3.386827 0.004975124 0.606965174 1.693292e-05 1.456530e-06 0.01419569 0.73859618 0.003691376 0.0003116975
#path:hsa05110  21.28218  84.18973    5.4202626      5.922305 0.004975124 0.004975124 1.122736e-04 8.600873e-06 0.01419569 0.02871357 0.012237826 0.0009202934
#path:hsa04145   0.00000 102.93799           NA      5.303335 0.004975124          NA 2.424942e-04 1.764759e-05 0.01419569         NA 0.016599259 0.0012588614
#path:hsa05100  67.58309 132.08113    4.2177527      5.750119 0.004975124 0.004975124 4.586545e-04 3.192445e-05 0.01419569 0.02871357 0.016599259 0.0017079580
#path:hsa05016   4.18750 123.28851   -0.1721632      3.961277 0.004975124 0.880597015 6.852905e-04 4.633032e-05 0.01419569 0.94116718 0.016599259 0.0019829376
#path:hsa05152 140.10374 233.84075    6.4536017      7.085604 0.004975124 0.004975124 1.082127e-03 7.069969e-05 0.01419569 0.02871357 0.021445796 0.0022802073


#Show clearer results (summarized form)
head(Summary(peRes, pathNames = kpn, totalAcc = FALSE, totalPert = FALSE, pAcc = FALSE, pORA = FALSE, comb.pv = NULL, order.by = "pPert"))
#Results
#                                           pathNames       pPert  pPert.fdr
#path:hsa03013                          RNA transport 0.004975124 0.01419569
#path:hsa04014                  Ras signaling pathway 0.004975124 0.01419569
#path:hsa04015                 Rap1 signaling pathway 0.004975124 0.01419569
#path:hsa04020              Calcium signaling pathway 0.004975124 0.01419569
#path:hsa04060 Cytokine-cytokine receptor interaction 0.004975124 0.01419569
#path:hsa04062            Chemokine signaling pathway 0.004975124 0.01419569



###########################################################################################################################################################

#Here starts the section of graphical representation
#Of course I won't be able to link the pictures here, i'll describe them as best as possible.

#plot function de R pour tout les résultats, c'est très moche.
plot(peRes)

#This plot shows pathway level statistics
plot(peRes, c("pAcc", "pORA"), comb.pv.func = compute.normalInv, threshold = .01)


#These plots can either be two way or boot type, no other type exists in this package
plot(peRes@pathways[["path:hsa05216"]], type = "two.way")
#Result
#This plot has two variables, log2FC as Y and Acc (accuracy) as x, it also has colored dots, doesn't seem to mean much

plot(peRes@pathways[["path:hsa05216"]], type = "boot")
#Result
#This plot is much better, it shows pathway level statistics: perturbation accumulation versus the measured expression change (above)
#and the bootstrap simulations of the perturbation accumulation (below)


#This set of functions allows the visualization of the propagation across the pathway
#It is all defined by the path:hsa05216, which is the 'thyroid cancer' pathway
dev.new()
pdf("Testing_pdf_output.pdf")

p <- peRes@pathways[["path:hsa05216"]]
g <- layoutGraph(p@map, layoutType = "dot")
graphRenderInfo(g) <- list(fixedsize = FALSE)
edgeRenderInfo(g) <- peEdgeRenderInfo(p)
nodeRenderInfo(g) <- peNodeRenderInfo(p)
renderGraph(g)

dev.off()
#Results
#Shows the perturbation propagation on the Thyroid cancer signaling pathway
#It's basically a type of network
#Currently the resolution is meh, but that can be adjusted with the png() function, among others


###########################################################################################################################################################


#If we wanted to do signalling pathway analysis, we would use the following function
#pDisRes <- pDis(x = fc, graphs = kpg, ref = ref, nboot = 200, verbose = FALSE)
#Results
#                   pathNames       ppDis ppDis.fdr
#path:hsa04976 Bile secretion 0.004975124 0.2437811
#path:hsa04390 Hippo signaling pathway 0.009950249 0.2437811
#path:hsa05030 Cocaine addiction 0.009950249 0.2437811
#path:hsa05142 Chagas disease (American trypanosomiasis) 0.009950249 0.2437811
#path:hsa05143 African trypanosomiasis 0.009950249 0.2437811
#path:hsa05031 Amphetamine addiction 0.014925373 0.2437811

